pubsub/subscription.go

```diff
@@ -1387,23 +1417,52 @@ func (s *Subscription) Receive(ctx context.Context, f func(context.Context, *Mes
 						// Return nil if the context is done, not err.
 						return nil
 					}
-					iter.eoMu.RLock()
-					msgAckHandler(msg, iter.enableExactlyOnceDelivery)
-					iter.eoMu.RUnlock()
+					if iter.enableTracing {
+						ccSpan.End()
+					}
 
 					wg.Add(1)
 					// Make sure the subscription has ordering enabled before adding to scheduler.
 					var key string
 					if s.enableOrdering {
 						key = msg.OrderingKey
 					}
+					// TODO(deklerk): Can we have a generic handler at the
+					// constructor level?
+					var schedulerSpan trace.Span
+					if iter.enableTracing {
+						_, schedulerSpan = startSpan(otelCtx, scheduleSpanName, "")
+					}
 					msgLen := len(msg.Data)
 					if err := sched.Add(key, msg, func(msg interface{}) {
+						m := msg.(*Message)
+						if iter.enableTracing {
+							schedulerSpan.End()
+						}
 						defer wg.Done()
-						defer fc.release(ctx, msgLen)
-						f(ctx2, msg.(*Message))
+						var ps trace.Span
+						if iter.enableTracing {
+							_, ps = startSpan(otelCtx, processSpanName, s.ID())
+							old := ackh.doneFunc
+							ackh.doneFunc = func(ackID string, ack bool, r *ipubsub.AckResult, receiveTime time.Time) {
+								var eventString string
+								if ack {
+									eventString = eventAckCalled
+								} else {
+									eventString = eventNackCalled
+								}
+								ps.AddEvent(eventString)
+								ps.SetAttributes(semconv.MessagingOperationProcess)
+								ps.End()
+								old(ackID, ack, r, receiveTime)
+							}
+						}
+						defer fc.release(otelCtx, msgLen)
+						f(ctx2, m)
```

Comment:
Could this function get sent a context from the `startSpan(otelCtx, processSpanName, s.ID())` ideally we'd have our handler get a context with a span that originated from the process span.